port/unix/omrsysinfo.c

```diff
@@ -3071,3 +3158,356 @@ omrsysinfo_os_kernel_info(struct OMRPortLibrary *portLibrary, struct OMROSKernel
+	strcpy(cgEntry->subsystem, subsystem);
+	cgEntry->cgroup = cgEntry->subsystem + strlen(subsystem) + 1;
+	strcpy(cgEntry->cgroup, cgroupName);
+
+	if (NULL == *cgEntryList) {
+		*cgEntryList = cgEntry;
+		cgEntry->next = cgEntry; /* create a circular list */
+	} else {
+		CgroupEntry *first = *cgEntryList;
+		*cgEntryList = cgEntry;
+		cgEntry->next = first->next;
+		first->next = cgEntry;
+	}
+
+_end:
+	return rc;
+}
+
+/**
+ * Reads /proc/<pid>/cgroup to get information about the cgroups to which the given
+ * process belongs.
+ *
+ * @param[in] portLibrary pointer to OMRPortLibrary
+ * @param[in] pid process id
+ * @param[out] cgroupEntryList pointer to CgroupEntry *. On successful return, *cgroupEntry
+ * points to a circular linked list. Each element of the list is populated based on the contents 
+ * of /proc/<pid>/cgroup file. 
+ *
+ * returns 0 on success, negative code on error
+ */
+static int32_t
+readCgroupFile(struct OMRPortLibrary *portLibrary, int pid, CgroupEntry **cgroupEntryList)
+{
+	char cgroup[PATH_MAX];
+	/* This array should be large enough to read names of all subsystems. 1024 should be enough. */
+	char subsystems[1024];
+	FILE *cgroupFile = NULL;
+	CgroupEntry *cgEntryList = NULL;
+	int32_t rc = 0;
+
+	Assert_PRT_true(NULL != cgroupEntryList);
+	
+	portLibrary->str_printf(portLibrary, cgroup, sizeof(cgroup), "/proc/%d/cgroup", pid);
+	cgroupFile = fopen(cgroup, "r");
+	if (NULL == cgroupFile) {
+		rc = portLibrary->error_set_last_error(portLibrary, errno, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_FOPEN_FAILED);
+		goto _end;
+	}
+
+	while (0 == feof(cgroupFile)) {
+		char *cursor = NULL;
+		char *separator = NULL;
+		int32_t hierId = -1;
+
+		/* Following is the description of /proc/<pid>/cgroup copied from 'man' page for proc:
+ 		 *
+ 		 * Each entry in /proc/<pid>/cgroup is of type:
+		 * 	5:cpuacct,cpu,cpuset:/daemons
+		 * The colon-separated fields are, from left to right:
+		 * 	1. hierarchy ID number
+		 * 	2. set of subsystems bound to the hierarchy
+		 * 	3. control group in the hierarchy to which the process belongs
+		 */
+		rc = fscanf(cgroupFile, "%d:%[^:]:%s", &hierId, subsystems, cgroup);
+		if (EOF == rc) {
+			rc = 0;
+			break;
+		} else if (3 != rc) {
+			rc = portLibrary->error_set_last_error_with_message(portLibrary, OMRPORT_ERROR_SYSINFO_PROCESS_CGROUP_FILE_READ_FAILED, "unexpcted format of process cgroup file");
+			goto _end;
+		}
+		cursor = subsystems;
+		do {
+			const char *cgName = NULL;
+
+			separator = strchr(cursor, ',');
+			if (NULL != separator) {
+				*separator = '\0';
+			}
+
+			cgName = getCgroupNameForSubsystem(portLibrary, cgEntryList, cursor);
+
+			if (NULL == cgName) {
+				rc = addCgroupEntry(portLibrary, &cgEntryList, hierId, cursor, cgroup);
+				if (0 != rc) {
+					goto _end;
+				}
+			} else {
+				char errBuf[512];
+
+				portLibrary->str_printf(portLibrary, errBuf, sizeof(errBuf), "multiple entries for subsystem %s found in process cgroup file", cursor);
```

Comment:
Truncation concern?  See review comments below for more details. (I started reviewing at the bottom)